bridge/gitlab/import.go

```diff
@@ -0,0 +1,351 @@
+	if err != nil {
+		return nil, err
+	}
+
+	// resolve bug
+	b, err := repo.ResolveBugCreateMetadata(keyGitlabUrl, issue.WebURL)
+	if err != nil && err != bug.ErrBugNotExist {
+		return nil, err
+	}
+
+	// if bug was never imported
+	if err == bug.ErrBugNotExist {
+		cleanText, err := text.Cleanup(issue.Description)
+		if err != nil {
+			return nil, err
+		}
+
+		// create bug
+		b, _, err = repo.NewBugRaw(
+			author,
+			issue.CreatedAt.Unix(),
+			issue.Title,
+			cleanText,
+			nil,
+			map[string]string{
+				keyOrigin:    target,
+				keyGitlabId:  parseID(issue.ID),
+				keyGitlabUrl: issue.WebURL,
+			},
+		)
+
+		if err != nil {
+			return nil, err
+		}
+
+		// importing a new bug
+		gi.importedIssues++
+
+		return b, nil
+	}
+
+	return b, nil
+}
+
+func (gi *gitlabImporter) ensureNote(repo *cache.RepoCache, b *cache.BugCache, note *gitlab.Note) error {
+	id := parseID(note.ID)
+
+	// ensure issue author
+	author, err := gi.ensurePerson(repo, note.Author.ID)
+	if err != nil {
+		return err
+	}
+
+	hash, errResolve := b.ResolveOperationWithMetadata(keyGitlabId, id)
+	if errResolve != cache.ErrNoMatchingOp {
+		return err
+	}
+
+	noteType, body := GetNoteType(note)
+	switch noteType {
+	case NOTE_CLOSED:
+		_, err = b.CloseRaw(
+			author,
+			note.CreatedAt.Unix(),
+			map[string]string{
+				keyGitlabId:  id,
+				keyGitlabUrl: "",
+			},
+		)
+		return err
+
+	case NOTE_REOPENED:
+		_, err = b.OpenRaw(
+			author,
+			note.CreatedAt.Unix(),
+			map[string]string{
+				keyGitlabId:  id,
+				keyGitlabUrl: "",
+			},
+		)
+		return err
+
+	case NOTE_DESCRIPTION_CHANGED:
+		issue := gi.iterator.IssueValue()
+
+		// since gitlab doesn't provide the issue history
+		// we should check for "changed the description" notes and compare issue texts
+		// TODO: Check only one time and ignore next 'description change' within one issue
+		if issue.Description != b.Snapshot().Comments[0].Message {
+
+			// comment edition
+			_, err = b.EditCommentRaw(
+				author,
+				note.UpdatedAt.Unix(),
+				target,
+				issue.Description,
+				map[string]string{
+					keyGitlabId:  id,
+					keyGitlabUrl: "",
+				},
+			)
+
+			return err
+
+		}
+
+	case NOTE_COMMENT:
+
+		cleanText, err := text.Cleanup(body)
+		if err != nil {
+			return err
+		}
+
+		// if we didn't import the comment
+		if errResolve == cache.ErrNoMatchingOp {
+
+			// add comment operation
+			_, err = b.AddCommentRaw(
+				author,
+				note.CreatedAt.Unix(),
+				cleanText,
+				nil,
+				map[string]string{
+					keyGitlabId:  id,
+					keyGitlabUrl: "",
+				},
+			)
+
+			return err
+		}
+
+		// if comment was already exported
+
+		// if note wasn't updated
+		if note.UpdatedAt.Equal(*note.CreatedAt) {
+			return nil
+		}
+
+		// search for last comment update
+		comment, err := b.Snapshot().SearchComment(hash)
+		if err != nil {
+			return err
+		}
+
+		// compare local bug comment with the new note body
+		if comment.Message != cleanText {
```

Comment:
I thing you have a problem here, what happen if you have, say, 2 comment editions, import, add more edition and import again ?

Reply:
The latest edition will be imported and added as comment edition.

Reply:
If the comment was never imported the latest version of the comment will be added as new comment locally